Add support for float[]? arguments in native_functions.yaml

dreiss opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37177 Update interpolate to use new upsample overloads
* #37176 Add interpolate-style overloads to aten::upsample* ops
* **#37175 Add support for float[]? arguments in native_functions.yaml**





Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994/)

Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994)

dreiss committed:
Add support for float[]? arguments in native_functions.yaml



Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994/)

[ghstack-poisoned]

dreiss commented on torch/csrc/jit/frontend/tracer.cpp:
```diff
@@ -765,7 +765,41 @@ void addInputs(
 
 void addInputs(Node* n, const char* name, ArrayRef<double> value) {
-  AT_ERROR("Tracing float lists currently not supported!");
+  AT_ERROR("XXX");
+  // using ArgumentStash = jit::tracer::ArgumentStash;
+  // std::vector<Value*> info = ArgumentStash::hasDoubleArrayRef(name)
+  //     ? ArgumentStash::popDoubleArrayRef(name)
+  //     : ArgumentStash::DoubleArrayRefTrace(value.size());
+
+  // auto& g = getTracingState()->graph;
+  // for (size_t i = 0; i < info.size(); ++i) {
+  //   if (info[i] != nullptr)
+  //     continue;
+  //   info[i] = g->insertConstant(value[i]);
+  //   recordSourceLocation(info[i]->node());
+  // }
+  // for (jit::Value* v : info) {
+  //   if (*v->type() != *jit::FloatType::get()) {
+  //     throw std::runtime_error(
+  //         "Type mismatch in setposattr for ArrayRef<double>. Check that your program "
+  //         "is valid without tracing, and please file a bug report if it is.");
+  //   }
+  // }
+  // n->addInput(
+  //     g->insertNode(g->createList(jit::FloatType::get(), info))->output());
```
Can anyone give me some guidance on how to implement this?  Is ArgumentStash relevant here?  Do I need to mirror all of those `IntArrayRef` methods for doubles?

jamesr66a replied:
I don't think ArgumentStash is relevant. iirc the place where it's relevant is when we switch return values for things like `sizes()` to be Tensors during tracing so we can record them. Downstream, someone might use those Tensors in a list that specifies sizes or something, and because the int values contained in those Tensors are the results of `sizes()`, we want to record them because they can affect the flow of the program.

OTOH, I don't think any float-valued stuff participates in that special-case recording behavior, so I think you can omit that stuff here.

dreiss commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -345,6 +346,29 @@ inline c10::OptionalArray<int64_t> PythonArgs::intlistOptional(int i) {
+    return {};
+  }
+  PyObject* arg = args[i];
+  // XXX no auto-broadcast
```
`intlist` does automatic expansion from scalars to fixed-size lists.  We don't need that, right?

gchanan replied:
I think it's fine to skip for now -- IMO we don't know if we need it or not yet.

dreiss commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -345,6 +346,29 @@ inline c10::OptionalArray<int64_t> PythonArgs::intlistOptional(int i) {
+  for (int idx = 0; idx < size; idx++) {
+    PyObject* obj = tuple ? PyTuple_GET_ITEM(arg, idx) : PyList_GET_ITEM(arg, idx);
+    try {
+      // XXX no variable check
```
`intlist` checks for variables here.  Can we safely ignore that?

gchanan replied:
that looks to be some hack that has to do tracing, I don't know whether we need to apply that same hack or whether that's even relevant anymore.  @eellison do you know?

eellison replied:
@jamesr66a said he would take a look

gchanan commented on aten/src/ATen/native_parse.py:
```diff
@@ -78,6 +78,10 @@ def type_argument_translations(arg):
         t = 'double'
     elif t == 'float?':
         t = 'double?'
+    elif t == 'float[]':
```
oh, I think this is where you needed the "using" from the IntArrayRef? PR.  Better if you keep them together, but not a big deal.

dreiss committed:
Update on "Add support for float[]? arguments in native_functions.yaml"






Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994/)

Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994)

[ghstack-poisoned]

dreiss committed:
Update on "Add support for float[]? arguments in native_functions.yaml"






Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994/)

Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994)

[ghstack-poisoned]

dreiss commented:
This is cleaned, tested, and ready for review.

dreiss commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -162,6 +162,8 @@ def save_arg(arg, is_output):
             saved_variables.append('std::vector<int64_t> {};'.format(name))
         elif arg['type'] == 'c10::optional<IntArrayRef>':
             saved_variables.append('c10::OptionalArray<int64_t> {};'.format(name))
+        elif arg['type'] == 'c10::optional<ArrayRef<double>> ':
```
I will bet my job that none of you reviewers would have caught the extra space in this string literal, which makes this clause totally ineffective.  You know who caught it?  ASAN caught it.

#thanksASAN

dreiss committed:
Update on "Add support for float[]? arguments in native_functions.yaml"






Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994/)

Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994)

[ghstack-poisoned]

dreiss committed:
Update on "Add support for float[]? arguments in native_functions.yaml"






Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994/)

Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994)

[ghstack-poisoned]

dreiss committed:
Update on "Add support for float[]? arguments in native_functions.yaml"






Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994/)

Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994)

[ghstack-poisoned]

dreiss committed:
Update on "Add support for float[]? arguments in native_functions.yaml"






Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994/)

Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994)

[ghstack-poisoned]

gchanan reviewed (approved):
what happens if you try to use a non-optional `float[]`?  You get an error that some python map doesn't have an entry for that?

dreiss commented:
> what happens if you try to use a non-optional `float[]`? You get an error that some python map doesn't have an entry for that?

Build error at code-gen time
```
RuntimeError: type 'ArrayRef<double>' is not supported
```

dreiss committed:
Update on "Add support for float[]? arguments in native_functions.yaml"






Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994/)

Differential Revision: [D21209994](https://our.internmc.facebook.com/intern/diff/D21209994)

[ghstack-poisoned]
